modules/gdpr_fields/src/EntityTraversal.php

```diff
@@ -0,0 +1,338 @@
+<?php
+
+namespace Drupal\gdpr_fields;
+
+use Drupal\Core\Entity\EntityFieldManagerInterface;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Entity\FieldableEntityInterface;
+use Drupal\gdpr_fields\Entity\GdprField;
+use Drupal\gdpr_fields\Entity\GdprFieldConfigEntity;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Base class for traversing entities.
+ *
+ * @package Drupal\gdpr_fields
+ */
+abstract class EntityTraversal implements EntityTraversalInterface {
+
+  /**
+   * Entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * Entity field manager.
+   *
+   * @var \Drupal\Core\Entity\EntityFieldManagerInterface
+   */
+  protected $entityFieldManager;
+
+  /**
+   * Entity storage for GDPR config entities.
+   *
+   * @var \Drupal\Core\Entity\EntityStorageInterface
+   */
+  protected $configStorage;
+
+  /**
+   * Reverse relationship information.
+   *
+   * @var \Drupal\gdpr_fields\Entity\GdprField[]
+   */
+  private $reverseRelationshipFields = NULL;
+
+  /**
+   * The starting entity for the traversal.
+   *
+   * @var \Drupal\Core\Entity\EntityInterface
+   */
+  protected $baseEntity;
+
+  /**
+   * Whether or not the traversal has happened successfully.
+   *
+   * @var bool
+   */
+  protected $success = NULL;
+
+  /**
+   * The processed entities.
+   *
+   * @var \Drupal\Core\Entity\EntityInterface[]
+   */
+  protected $entities = [];
+
+  /**
+   * The results of the traversal.
+   *
+   * @var array
+   */
+  protected $results = [];
+
+  /**
+   * EntityTraversal constructor.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager.
+   * @param \Drupal\Core\Entity\EntityFieldManagerInterface $entityFieldManager
+   *   The entity field manager.
+   * @param \Drupal\Core\Entity\EntityInterface $base_entity
+   *   The starting entity for the traversal.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   */
+  public function __construct(EntityTypeManagerInterface $entityTypeManager, EntityFieldManagerInterface $entityFieldManager, EntityInterface $base_entity) {
+    $this->entityTypeManager = $entityTypeManager;
+    $this->entityFieldManager = $entityFieldManager;
+    $this->configStorage = $this->entityTypeManager->getStorage('gdpr_fields_config');
+    $this->baseEntity = $base_entity;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, EntityInterface $base_entity) {
+    return new static(
+      $container->get('entity_type.manager'),
+      $container->get('entity_field.manager'),
+      $base_entity
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function traverse() {
+    if (is_null($this->success)) {
```

Comment:
Do we want this check here? This means you can't ever call `traverse` twice on the same entity. If there is ever a situation where you want to traverse, then modify the entity based on results, then traverse again then that won't work.

Reply:
I'm not sure we want multiple traversals. If we did we could provide a default `$force = FALSE` parameter. @andrewbelcher ?

Reply:
I think if you want to re-do it you can instantiate a new traverser - all the getter methods call `$this->traverse()` to ensure the data has been processed.